velox/common/encode/Base64.cpp

```diff
@@ -324,65 +327,62 @@ uint8_t Base64::Base64ReverseLookup(
 
 size_t
 Base64::decode(const char* src, size_t src_len, char* dst, size_t dst_len) {
-  return decodeImpl(src, src_len, dst, dst_len, kBase64ReverseIndexTable, true);
+  return decodeImpl(src, src_len, dst, dst_len, kBase64ReverseIndexTable);
 }
 
 // static
-size_t
-Base64::calculateDecodedSize(const char* data, size_t& size, bool withPadding) {
+size_t Base64::calculateDecodedSize(const char* data, size_t& size) {
   if (size == 0) {
     return 0;
   }
 
-  auto needed = (size / 4) * 3;
-  if (withPadding) {
-    // If the pad characters are included then the source string must be a
-    // multiple of 4 and we can query the end of the string to see how much
-    // padding exists.
-    if (size % 4 != 0) {
+  // Check if the input data is padded
+  if (isPadded(data, size)) {
+    // If padded, ensure that the string length is a multiple of the encoded
+    // block size
+    if (size % kEncodedBlockSize != 0) {
       throw Base64Exception(
           "Base64::decode() - invalid input string: "
-          "string length is not multiple of 4.");
+          "string length is not a multiple of 4.");
     }
 
+    auto needed = (size * kBinaryBlockSize) / kEncodedBlockSize;
     auto padding = countPadding(data, size);
     size -= padding;
-    return needed - padding;
+
+    // Adjust the needed size for padding
+    return needed -
```

Comment:
nit: I haven't had the opportunity to review the subsequent PRs, but my initial confusion during the review was due to the intertwining of some code that is specific to base64 and some that applies to both 64 and 32. I had to delve into the details of both to assure myself of their functionality.  It would be great if you could annotate/comment on some of these complex calculations in the final version (where this code is reused) to clarify how we arrived at these formulas. This would greatly aid future readers. Thanks

Reply:
I have updated the comments regarding the need for the calculations we have done here. 
The calculation is a bit different in base32 because of the need to handle some exceptional cases in handling padded lengths.